Flaky test init TestOraclePoolIntegrationSmokeTest

```
2020-07-24T15:47:18.1678486Z [ERROR] Tests run: 256, Failures: 2, Errors: 0, Skipped: 203, Time elapsed: 486.597 s <<< FAILURE! - in TestSuite
2020-07-24T15:47:18.1685135Z [ERROR] init(io.prestosql.plugin.oracle.TestOraclePoolIntegrationSmokeTest)  Time elapsed: 255.865 s  <<< FAILURE!
2020-07-24T15:47:18.1691519Z java.lang.RuntimeException: java.sql.SQLRecoverableException: IO Error: Connection reset by peer, Authentication lapse 0 ms.
2020-07-24T15:47:18.1696116Z 	at io.prestosql.plugin.oracle.TestingOracleServer.<init>(TestingOracleServer.java:75)
2020-07-24T15:47:18.1702161Z 	at io.prestosql.plugin.oracle.BaseOracleIntegrationSmokeTest.createQueryRunner(BaseOracleIntegrationSmokeTest.java:41)
2020-07-24T15:47:18.1707186Z 	at io.prestosql.testing.AbstractTestQueryFramework.init(AbstractTestQueryFramework.java:80)
2020-07-24T15:47:18.1712382Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-07-24T15:47:18.1717673Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-07-24T15:47:18.1726507Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-07-24T15:47:18.1731457Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-07-24T15:47:18.1737032Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-07-24T15:47:18.1741666Z 	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
2020-07-24T15:47:18.1746241Z 	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:217)
2020-07-24T15:47:18.1751062Z 	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:144)
2020-07-24T15:47:18.1756226Z 	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:169)
2020-07-24T15:47:18.1760462Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
2020-07-24T15:47:18.1767100Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-07-24T15:47:18.1773300Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-07-24T15:47:18.1776512Z 	at java.base/java.lang.Thread.run(Thread.java:834)
2020-07-24T15:47:18.1782769Z Caused by: java.sql.SQLRecoverableException: IO Error: Connection reset by peer, Authentication lapse 0 ms.
2020-07-24T15:47:18.1786751Z 	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:874)
2020-07-24T15:47:18.1791522Z 	at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:793)
2020-07-24T15:47:18.1796885Z 	at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:57)
2020-07-24T15:47:18.1801027Z 	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:747)
2020-07-24T15:47:18.1805352Z 	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:562)
2020-07-24T15:47:18.1809592Z 	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:677)
2020-07-24T15:47:18.1813840Z 	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:228)
2020-07-24T15:47:18.1819013Z 	at io.prestosql.plugin.oracle.TestingOracleServer.<init>(TestingOracleServer.java:67)
2020-07-24T15:47:18.1820414Z 	... 15 more
2020-07-24T15:47:18.1825327Z Caused by: java.io.IOException: Connection reset by peer, Authentication lapse 0 ms.
2020-07-24T15:47:18.1829643Z 	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:870)
2020-07-24T15:47:18.1830870Z 	... 22 more
2020-07-24T15:47:18.1834453Z Caused by: java.io.IOException: Connection reset by peer
2020-07-24T15:47:18.1838446Z 	at java.base/sun.nio.ch.FileDispatcherImpl.read0(Native Method)
2020-07-24T15:47:18.1842564Z 	at java.base/sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
2020-07-24T15:47:18.1846741Z 	at java.base/sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:276)
2020-07-24T15:47:18.1850231Z 	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:245)
2020-07-24T15:47:18.1853368Z 	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:223)
2020-07-24T15:47:18.1858174Z 	at java.base/sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:355)
2020-07-24T15:47:18.1862214Z 	at oracle.net.nt.TimeoutSocketChannel.read(TimeoutSocketChannel.java:174)
2020-07-24T15:47:18.1866401Z 	at oracle.net.ns.NSProtocolNIO.doSocketRead(NSProtocolNIO.java:555)
2020-07-24T15:47:18.1870358Z 	at oracle.net.ns.NIOPacket.readNIOPacket(NIOPacket.java:403)
2020-07-24T15:47:18.1874631Z 	at oracle.net.ns.NSProtocolNIO.negotiateConnection(NSProtocolNIO.java:127)
2020-07-24T15:47:18.1880071Z 	at oracle.net.ns.NSProtocol.connect(NSProtocol.java:340)
2020-07-24T15:47:18.1883988Z 	at oracle.jdbc.driver.T4CConnection.connect(T4CConnection.java:1596)
2020-07-24T15:47:18.1887741Z 	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:588)
2020-07-24T15:47:18.1889114Z 	... 22 more
```

https://github.com/prestosql/presto/runs/907209390

This is very similar to https://github.com/prestosql/presto/issues/4654, both fails on opening connection. And it looks like we could retry that.


I got different outcome of the same issue:
```
020-08-21T13:18:03.4340593Z java.lang.RuntimeException: java.sql.SQLRecoverableException: IO Error: Connection reset by peer, Authentication lapse 0 ms.
2020-08-21T13:18:03.4340767Z 	at io.prestosql.plugin.oracle.TestingOracleServer.<init>(TestingOracleServer.java:75)
2020-08-21T13:18:03.4340939Z 	at io.prestosql.plugin.oracle.BaseOracleIntegrationSmokeTest.createQueryRunner(BaseOracleIntegrationSmokeTest.java:41)
2020-08-21T13:18:03.4341309Z 	at io.prestosql.testing.AbstractTestQueryFramework.init(AbstractTestQueryFramework.java:81)
2020-08-21T13:18:03.4341716Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-08-21T13:18:03.4341889Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-08-21T13:18:03.4342074Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-08-21T13:18:03.4342435Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-08-21T13:18:03.4342914Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-08-21T13:18:03.4347443Z 	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
2020-08-21T13:18:03.4351345Z 	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:217)
2020-08-21T13:18:03.4355559Z 	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:144)
2020-08-21T13:18:03.4360988Z 	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:169)
2020-08-21T13:18:03.4365306Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
2020-08-21T13:18:03.4370705Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-08-21T13:18:03.4375976Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-08-21T13:18:03.4378968Z 	at java.base/java.lang.Thread.run(Thread.java:834)
2020-08-21T13:18:03.4385487Z Caused by: java.sql.SQLRecoverableException: IO Error: Connection reset by peer, Authentication lapse 0 ms.
2020-08-21T13:18:03.4421877Z 	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:874)
2020-08-21T13:18:03.4423011Z 	at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:793)
2020-08-21T13:18:03.4423936Z 	at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:57)
2020-08-21T13:18:03.4424745Z 	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:747)
2020-08-21T13:18:03.4425557Z 	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:562)
2020-08-21T13:18:03.4426343Z 	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:677)
2020-08-21T13:18:03.4427133Z 	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:228)
2020-08-21T13:18:03.4427920Z 	at io.prestosql.plugin.oracle.TestingOracleServer.<init>(TestingOracleServer.java:67)
2020-08-21T13:18:03.4428924Z 	... 15 more
2020-08-21T13:18:03.4429183Z Caused by: java.io.IOException: Connection reset by peer, Authentication lapse 0 ms.
2020-08-21T13:18:03.4429443Z 	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:870)
2020-08-21T13:18:03.4429671Z 	... 22 more
2020-08-21T13:18:03.4429907Z Caused by: java.io.IOException: Connection reset by peer
2020-08-21T13:18:03.4430149Z 	at java.base/sun.nio.ch.FileDispatcherImpl.read0(Native Method)
2020-08-21T13:18:03.4430395Z 	at java.base/sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
2020-08-21T13:18:03.4430643Z 	at java.base/sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:276)
2020-08-21T13:18:03.4430895Z 	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:245)
2020-08-21T13:18:03.4431129Z 	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:223)
2020-08-21T13:18:03.4431370Z 	at java.base/sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:355)
2020-08-21T13:18:03.4431622Z 	at oracle.net.nt.TimeoutSocketChannel.read(TimeoutSocketChannel.java:174)
2020-08-21T13:18:03.4431977Z 	at oracle.net.ns.NSProtocolNIO.doSocketRead(NSProtocolNIO.java:555)
2020-08-21T13:18:03.4432219Z 	at oracle.net.ns.NIOPacket.readNIOPacket(NIOPacket.java:403)
2020-08-21T13:18:03.4432440Z 	at oracle.net.ns.NSProtocolNIO.negotiateConnection(NSProtocolNIO.java:127)
2020-08-21T13:18:03.4432663Z 	at oracle.net.ns.NSProtocol.connect(NSProtocol.java:340)
2020-08-21T13:18:03.4432866Z 	at oracle.jdbc.driver.T4CConnection.connect(T4CConnection.java:1596)
2020-08-21T13:18:03.4433092Z 	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:588)
2020-08-21T13:18:03.4433395Z 	... 22 more
2020-08-21T13:18:03.4433560Z 
2020-08-21T13:18:03.4433787Z [ERROR] destroy(io.prestosql.plugin.oracle.TestOraclePoolIntegrationSmokeTest)  Time elapsed: 0 s  <<< FAILURE!
2020-08-21T13:18:03.4434016Z java.lang.NullPointerException
2020-08-21T13:18:03.4434243Z 	at io.prestosql.plugin.oracle.BaseOracleIntegrationSmokeTest.destroy(BaseOracleIntegrationSmokeTest.java:51)
2020-08-21T13:18:03.4434492Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-08-21T13:18:03.4434740Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-08-21T13:18:03.4434993Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-08-21T13:18:03.4435231Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-08-21T13:18:03.4435464Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-08-21T13:18:03.4435685Z 	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
2020-08-21T13:18:03.4435920Z 	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:217)
2020-08-21T13:18:03.4436141Z 	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:144)
2020-08-21T13:18:03.4436368Z 	at org.testng.internal.TestMethodWorker.invokeAfterClassMethods(TestMethodWorker.java:217)
2020-08-21T13:18:03.4436604Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:115)
2020-08-21T13:18:03.4436839Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-08-21T13:18:03.4437303Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-08-21T13:18:03.4462631Z 	at java.base/java.lang.Thread.run(Thread.java:834)
2020-08-21T13:18:03.4462921Z 
2020-08-21T13:18:03.4546229Z 2020-08-21T08:18:03.454-0500 INFO JVM is shutting down, cleaning up
2020-08-21T13:18:03.4548307Z 2020-08-21T08:18:03.454-0500 INFO Life cycle stopping...
2020-08-21T13:18:03.4550338Z 2020-08-21T08:18:03.454-0500 INFO Life cycle stopped
2020-08-21T13:18:03.8814028Z [INFO] 
2020-08-21T13:18:03.8814600Z [INFO] Results:
2020-08-21T13:18:03.8814815Z [INFO] 
```

Yet another occurence:
```
2021-07-07T08:49:07.8729311Z [ERROR] init(io.trino.plugin.oracle.TestOraclePoolRemarksReportingConnectorSmokeTest)  Time elapsed: 248.107 s  <<< FAILURE!
2021-07-07T08:49:07.8736490Z java.lang.RuntimeException: java.sql.SQLRecoverableException: IO Error: Connection reset by peer, Authentication lapse 0 ms.
2021-07-07T08:49:07.8743889Z 	at io.trino.plugin.oracle.TestingOracleServer.<init>(TestingOracleServer.java:81)
2021-07-07T08:49:07.8754709Z 	at io.trino.plugin.oracle.TestOraclePoolRemarksReportingConnectorSmokeTest.createQueryRunner(TestOraclePoolRemarksReportingConnectorSmokeTest.java:38)
2021-07-07T08:49:07.8762966Z 	at io.trino.testing.AbstractTestQueryFramework.init(AbstractTestQueryFramework.java:91)
2021-07-07T08:49:07.8769893Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2021-07-07T08:49:07.8776721Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2021-07-07T08:49:07.8787855Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2021-07-07T08:49:07.8802144Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2021-07-07T08:49:07.8808366Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2021-07-07T08:49:07.8819363Z 	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
2021-07-07T08:49:07.8824062Z 	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:217)
2021-07-07T08:49:07.8829110Z 	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:144)
2021-07-07T08:49:07.8840176Z 	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:169)
2021-07-07T08:49:07.8845809Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
2021-07-07T08:49:07.8856323Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2021-07-07T08:49:07.8861451Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2021-07-07T08:49:07.8865207Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2021-07-07T08:49:07.8902724Z Caused by: java.sql.SQLRecoverableException: IO Error: Connection reset by peer, Authentication lapse 0 ms.
2021-07-07T08:49:07.8904312Z 	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:874)
2021-07-07T08:49:07.8905926Z 	at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:793)
2021-07-07T08:49:07.8908802Z 	at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:57)
2021-07-07T08:49:07.8910727Z 	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:747)
2021-07-07T08:49:07.8912143Z 	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:562)
2021-07-07T08:49:07.8914156Z 	at io.trino.plugin.jdbc.DriverConnectionFactory.openConnection(DriverConnectionFactory.java:65)
2021-07-07T08:49:07.8916695Z 	at io.trino.plugin.jdbc.RetryingConnectionFactory.lambda$openConnection$0(RetryingConnectionFactory.java:54)
2021-07-07T08:49:07.8918955Z 	at net.jodah.failsafe.Functions.lambda$get$0(Functions.java:48)
2021-07-07T08:49:07.8920498Z 	at net.jodah.failsafe.RetryPolicyExecutor.lambda$supply$0(RetryPolicyExecutor.java:62)
2021-07-07T08:49:07.8922135Z 	at net.jodah.failsafe.Execution.executeSync(Execution.java:129)
2021-07-07T08:49:07.8923660Z 	at net.jodah.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
2021-07-07T08:49:07.8925290Z 	at net.jodah.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:67)
2021-07-07T08:49:07.8927502Z 	at io.trino.plugin.jdbc.RetryingConnectionFactory.openConnection(RetryingConnectionFactory.java:54)
2021-07-07T08:49:07.8931956Z 	at io.trino.plugin.oracle.TestingOracleServer.<init>(TestingOracleServer.java:73)
2021-07-07T08:49:07.8934115Z 	... 15 more
2021-07-07T08:49:07.8937534Z Caused by: java.io.IOException: Connection reset by peer, Authentication lapse 0 ms.
2021-07-07T08:49:07.8941190Z 	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:870)
2021-07-07T08:49:07.8944178Z 	... 28 more
2021-07-07T08:49:07.8949254Z Caused by: java.io.IOException: Connection reset by peer
2021-07-07T08:49:07.8952795Z 	at java.base/sun.nio.ch.FileDispatcherImpl.read0(Native Method)
2021-07-07T08:49:07.8958211Z 	at java.base/sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
2021-07-07T08:49:07.8961839Z 	at java.base/sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:276)
2021-07-07T08:49:07.8967892Z 	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:245)
2021-07-07T08:49:07.8968716Z 	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:223)
2021-07-07T08:49:07.8971450Z 	at java.base/sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:353)
2021-07-07T08:49:07.8982128Z 	at oracle.net.nt.TimeoutSocketChannel.read(TimeoutSocketChannel.java:174)
2021-07-07T08:49:07.8988692Z 	at oracle.net.ns.NSProtocolNIO.doSocketRead(NSProtocolNIO.java:555)
2021-07-07T08:49:07.8990197Z 	at oracle.net.ns.NIOPacket.readNIOPacket(NIOPacket.java:403)
2021-07-07T08:49:07.8991740Z 	at oracle.net.ns.NSProtocolNIO.negotiateConnection(NSProtocolNIO.java:127)
2021-07-07T08:49:07.8993202Z 	at oracle.net.ns.NSProtocol.connect(NSProtocol.java:340)
2021-07-07T08:49:07.8995051Z 	at oracle.jdbc.driver.T4CConnection.connect(T4CConnection.java:1596)
2021-07-07T08:49:07.9002029Z 	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:588)
2021-07-07T08:49:07.9003055Z 	... 28 more
2021-07-07T08:49:07.9003279Z 
2021-07-07T08:49:08.4682584Z [INFO] 
2021-07-07T08:49:08.4685460Z [INFO] Results:
2021-07-07T08:49:08.4687389Z [INFO] 
2021-07-07T08:49:08.4689902Z [ERROR] Failures: 
2021-07-07T08:49:08.4734033Z [ERROR]   TestOraclePoolRemarksReportingConnectorSmokeTest>AbstractTestQueryFramework.init:91->createQueryRunner:38 » Runtime
2021-07-07T08:49:08.4735917Z [INFO] 
2021-07-07T08:49:08.4736587Z [ERROR] Tests run: 245, Failures: 1, Errors: 0, Skipped: 105
2021-07-07T08:49:08.4737112Z [INFO] 
```

--- a/plugin/trino-oracle/pom.xml
+++ b/plugin/trino-oracle/pom.xml
@@ -76,6 +76,11 @@
         </dependency>
 
         <dependency>
+            <groupId>net.jodah</groupId>
+            <artifactId>failsafe</artifactId>
+        </dependency>
+
+        <dependency>
             <groupId>io.airlift</groupId>
             <artifactId>log-manager</artifactId>
             <scope>runtime</scope>

--- a/plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestingOracleServer.java
+++ b/plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestingOracleServer.java
@@ -13,43 +13,75 @@
  */
 package io.trino.plugin.oracle;
 
+import io.airlift.log.Logger;
 import io.trino.plugin.jdbc.BaseJdbcConfig;
 import io.trino.plugin.jdbc.ConnectionFactory;
 import io.trino.plugin.jdbc.DriverConnectionFactory;
 import io.trino.plugin.jdbc.RetryingConnectionFactory;
 import io.trino.plugin.jdbc.credential.StaticCredentialProvider;
+import net.jodah.failsafe.Failsafe;
+import net.jodah.failsafe.RetryPolicy;
 import oracle.jdbc.OracleDriver;
 import org.testcontainers.containers.OracleContainer;
 import org.testcontainers.utility.MountableFile;
 
 import java.io.Closeable;
-import java.io.IOException;
 import java.sql.Connection;
 import java.sql.SQLException;
 import java.sql.Statement;
+import java.time.temporal.ChronoUnit;
 
+import static io.airlift.testing.Closeables.closeAllSuppress;
 import static io.trino.testing.TestingConnectorSession.SESSION;
 import static java.lang.String.format;
 
 public class TestingOracleServer
-        extends OracleContainer
         implements Closeable
 {
+    private static final Logger log = Logger.get(TestingOracleServer.class);
+
+    private static final RetryPolicy<Object> CONTAINER_RETRY_POLICY = new RetryPolicy<>()
+            .withBackoff(1, 5, ChronoUnit.SECONDS)
+            .withMaxAttempts(5)
+            .onRetry(event -> log.warn(
+                    "Container initialization failed on attempt %s, will retry. Exception: %s",
+                    event.getAttemptCount(),
+                    event.getLastFailure().getMessage()));
+
     private static final String TEST_TABLESPACE = "trino_test";
 
     public static final String TEST_USER = "trino_test";
     public static final String TEST_SCHEMA = TEST_USER; // schema and user is the same thing in Oracle
     public static final String TEST_PASS = "trino_test_password";
 
+    private final TestingOracleContainer container;
+
     public TestingOracleServer()
     {
-        super("wnameless/oracle-xe-11g-r2");
+        container = Failsafe.with(CONTAINER_RETRY_POLICY).get(this::createContainer);
+    }
 
-        withCopyFileToContainer(MountableFile.forClasspathResource("init.sql"), "/docker-entrypoint-initdb.d/init.sql");
+    private TestingOracleContainer createContainer()
+    {
+        TestingOracleContainer container = new TestingOracleContainer("wnameless/oracle-xe-11g-r2");
+        container.withCopyFileToContainer(MountableFile.forClasspathResource("init.sql"), "/docker-entrypoint-initdb.d/init.sql");
 
-        start();
+        container.start();
+        try {
+            setUpDatabase(container);
+        }
+        catch (Exception e) {
+            closeAllSuppress(e, container::stop);
+            throw new RuntimeException(e);
+        }
 
-        try (Connection connection = getConnectionFactory().openConnection(SESSION);
+        return container;
+    }
+
+    private void setUpDatabase(TestingOracleContainer container)
+            throws Exception
+    {
+        try (Connection connection = getConnectionFactory(container).openConnection(SESSION);
                 Statement statement = connection.createStatement()) {
             // this is added to allow more processes on database, otherwise the tests end up giving
             // ORA-12519, TNS:no appropriate service handler found
@@ -58,34 +90,22 @@ public class TestingOracleServer
             statement.execute("ALTER SYSTEM SET processes=1000 SCOPE=SPFILE");
             statement.execute("ALTER SYSTEM SET disk_asynch_io = FALSE SCOPE = SPFILE");
         }
-        catch (SQLException e) {
-            throw new RuntimeException(e);
-        }
 
-        try {
-            execInContainer("/bin/bash", "/etc/init.d/oracle-xe", "restart");
-        }
-        catch (InterruptedException | IOException e) {
-            throw new RuntimeException(e);
-        }
+        container.execInContainer("/bin/bash", "/etc/init.d/oracle-xe", "restart");
 
-        waitUntilContainerStarted();
-        try (Connection connection = getConnectionFactory().openConnection(SESSION);
+        container.waitUntilContainerStarted();
+        try (Connection connection = getConnectionFactory(container).openConnection(SESSION);
                 Statement statement = connection.createStatement()) {
             statement.execute(format("CREATE TABLESPACE %s DATAFILE 'test_db.dat' SIZE 100M ONLINE", TEST_TABLESPACE));
             statement.execute(format("CREATE USER %s IDENTIFIED BY %s DEFAULT TABLESPACE %s", TEST_USER, TEST_PASS, TEST_TABLESPACE));
             statement.execute(format("GRANT UNLIMITED TABLESPACE TO %s", TEST_USER));
             statement.execute(format("GRANT ALL PRIVILEGES TO %s", TEST_USER));
         }
-        catch (SQLException e) {
-            throw new RuntimeException(e);
-        }
     }
 
-    @Override
     public String getJdbcUrl()
     {
-        return "jdbc:oracle:thin:@" + getHost() + ":" + getOraclePort() + ":" + getSid();
+        return container.getJdbcUrl();
     }
 
     public void execute(String sql)
@@ -95,7 +115,7 @@ public class TestingOracleServer
 
     public void execute(String sql, String user, String password)
     {
-        try (Connection connection = getConnectionFactory(user, password).openConnection(SESSION);
+        try (Connection connection = getConnectionFactory(getJdbcUrl(), user, password).openConnection(SESSION);
                 Statement statement = connection.createStatement()) {
             statement.execute(sql);
         }
@@ -104,16 +124,16 @@ public class TestingOracleServer
         }
     }
 
-    private ConnectionFactory getConnectionFactory()
+    private ConnectionFactory getConnectionFactory(TestingOracleContainer container)
     {
-        return getConnectionFactory(getUsername(), getPassword());
+        return getConnectionFactory(container.getJdbcUrl(), container.getUsername(), container.getPassword());
     }
 
-    private ConnectionFactory getConnectionFactory(String username, String password)
+    private ConnectionFactory getConnectionFactory(String connectionUrl, String username, String password)
     {
         DriverConnectionFactory connectionFactory = new DriverConnectionFactory(
                 new OracleDriver(),
-                new BaseJdbcConfig().setConnectionUrl(getJdbcUrl()),
+                new BaseJdbcConfig().setConnectionUrl(connectionUrl),
                 StaticCredentialProvider.of(username, password));
         return new RetryingConnectionFactory(connectionFactory);
     }
@@ -121,6 +141,21 @@ public class TestingOracleServer
     @Override
     public void close()
     {
-        stop();
+        container.stop();
+    }
+
+    private static class TestingOracleContainer
+            extends OracleContainer
+    {
+        public TestingOracleContainer(String dockerImageName)
+        {
+            super(dockerImageName);
+        }
+
+        @Override
+        protected void waitUntilContainerStarted()
+        {
+            super.waitUntilContainerStarted();
+        }
     }
 }

